### ISSUE TITLE
sql,kv: exponential queries can cause ranges to grow faster than they can split and crash a node

### ISSUE THREAD

**Issue by knz** [cockroachdb/cockroach#11208]
So I was exploring alternatives to #11203 and ran the following experiment:

First I create a single-node cluster with replication factor 1 (`echo 'num_replicas: 1' | ./cockroach set zone .default -f -`)

Then in the SQL shell:

```sql
create database t;
create table t.b(x bytes);
--- make a single 4MB row
insert into t.b(x) values(repeat(b'\x01', 4*1024*1024));
--- duplicate it a couple of times:
insert into t.b(x) select x from t.b; -- 1 more rows - 1 range total
insert into t.b(x) select x from t.b; -- 2 more rows - 1 range total
insert into t.b(x) select x from t.b; -- 4 more rows - 1 range total
insert into t.b(x) select x from t.b; -- 8 more rows - 1 range total
insert into t.b(x) select x from t.b; -- 16 more rows - 2 ranges total
-- (from this point each new insert doubles the number of ranges)

-- alternatively the following query creates ~80MB of data at a time, ie somewhat more than 1 range:
insert into t.b(x) select x from (select * from t.b limit 20); -- 20 more rows - +1 range
```

A couple of observations, which perhaps should translate into issues:
- if I run the inserts too fast behind each other, the node crashes with OOM (the ranges grow faster than they are split)
- the raftMu lock is being held for very long (seconds at  a time)
- the following error cropped up after ~200 rows:
```
W161123 18:37:17.126535 1888 kv/txn_coord_sender.go:771  [n1,split,s1,r22/1:/Table/51/1/196135{404…-535…}] heartbeat to "storage/replica_command.go:2446 (*Replica).adminSplitWithDescriptor" id=329f81be key=/Local/Range/"\xbb\x89\xfd\x02\xb8\xd0\x1c\x8e\xa7\x00\x01"/RangeDescriptor rw=true pri=0.06858520 iso=SE
RIALIZABLE stat=PENDING epo=0 ts=1479926225.099704909,0 orig=1479926225.099704909,0 max=1479926225.349704909,0 wto=false rop=false failed: failed to send RPC: sending to all 1 replicas failed; last error: range 22: replica {1 1 1} not lease holder; <nil> is
```
(I'm not sure that a `nil` lease holder is good news; remember this is a 1-node cluster)

cc @petermattis @spencerkimball

**Comment by knz** [comment]
replicaMu not happy either.

**Comment by spencerkimball** [comment]
The lease isn't getting refreshed so the node loses it. We could probably avoid that kind of error completely for 1 node clusters, but the optimization likely isn't worth the effort.

I think it's not OOMing because ranges are growing faster than they're split. Splitting, no matter how big the range is, will be incredibly fast as only metadata is altered. It's more likely just the size of your batches, which in this case will be 1G of data.

`raftMu` will be held for long periods because the size of the request to be applied gets large.

At some point, we may need to enforce limits on request sizes.

**Comment by knz** [comment]
Hmm even if I limit the batch sizes: `insert into t.b(x) select x from t.b limit 20` (20 rows, ~80MB of data) I still get OOM errors.

**Comment by spencerkimball** [comment]
Take a look at the heap profile...?

**Comment by knz** [comment]
yes will do

**Comment by knz** [comment]
Ah never mind I was writing the query incorrectly. You were right the culprit is the batch size.

**Comment by petermattis** [comment]
Large batches are problematic. I think we should document this as a known issue for 1.0.

**Comment by knz** [comment]
Fixed by  #15350.
